addons/website_sale/models/product_product.py

```diff
@@ -418,3 +418,69 @@ def _prepare_gmc_additional_info(self):
             gmc_info['item_group_id'] = self.product_tmpl_id.id
 
         return gmc_info
+    # ---------------------------------------------------------
+    # Product Ribbon Methods
+    # ---------------------------------------------------------
+
+    def _get_ribbon(self, product_prices):
+        """
+        Retrieve the appropriate ribbon for the product based on manual or automatic assignment.
+
+        This method first checks if a ribbon is manually set If no manual ribbon is set, it returns
+        appropriate ribbon based on the priority assigned in product.ribbon.
+
+        :param dict product_prices: A dictionary containing product pricing information.
+        :return: The ribbon that matches the product's criteria or an empty ribbon record.
+        :rtype: `product.ribbon` recordset.
+        """
+        product_ribbon_sudo = self.env['product.ribbon'].sudo()
+        if not self:
+            return product_ribbon_sudo
+
+        manually_set_ribbon = self.product_tmpl_id.website_ribbon_id or self.variant_ribbon_id
+        if manually_set_ribbon:
+            return manually_set_ribbon
+
+        auto_assign_ribbons = product_ribbon_sudo.search([('assign', '!=', 'manual')])
```

Comment:
This search will be made for every product on the `/shop` page. It would be best to avoid doing that and only do this search once per rendering. By default, there are 20 products on the `/shop` page so this research would be made 20 times except if some of the products displayed have a manually assigned ribbon.